File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-shared-5-4/PeriodicDates.cpp
// PeriodicDates.cpp 
#include "PeriodicDates.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

PeriodicDates::PeriodicDates(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //calculationStartDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* calculationStartDateNode = xmlNode->FirstChildElement("calculationStartDate");

   if(calculationStartDateNode){calculationStartDateIsNull_ = false;}
   else{calculationStartDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- calculationStartDateNode , address : " << calculationStartDateNode << std::endl;
   #endif
   if(calculationStartDateNode)
   {
      if(calculationStartDateNode->Attribute("href") || calculationStartDateNode->Attribute("id"))
      {
          if(calculationStartDateNode->Attribute("id"))
          {
             calculationStartDateIDRef_ = calculationStartDateNode->Attribute("id");
             calculationStartDate_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(calculationStartDateNode));
             calculationStartDate_->setID(calculationStartDateIDRef_);
             IDManager::instance().setID(calculationStartDateIDRef_,calculationStartDate_);
          }
          else if(calculationStartDateNode->Attribute("href")) { calculationStartDateIDRef_ = calculationStartDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { calculationStartDate_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(calculationStartDateNode));}
   }

   //calculationEndDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* calculationEndDateNode = xmlNode->FirstChildElement("calculationEndDate");

   if(calculationEndDateNode){calculationEndDateIsNull_ = false;}
   else{calculationEndDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- calculationEndDateNode , address : " << calculationEndDateNode << std::endl;
   #endif
   if(calculationEndDateNode)
   {
      if(calculationEndDateNode->Attribute("href") || calculationEndDateNode->Attribute("id"))
      {
          if(calculationEndDateNode->Attribute("id"))
          {
             calculationEndDateIDRef_ = calculationEndDateNode->Attribute("id");
             calculationEndDate_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(calculationEndDateNode));
             calculationEndDate_->setID(calculationEndDateIDRef_);
             IDManager::instance().setID(calculationEndDateIDRef_,calculationEndDate_);
          }
          else if(calculationEndDateNode->Attribute("href")) { calculationEndDateIDRef_ = calculationEndDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { calculationEndDate_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(calculationEndDateNode));}
   }

   //calculationPeriodFrequencyNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* calculationPeriodFrequencyNode = xmlNode->FirstChildElement("calculationPeriodFrequency");

   if(calculationPeriodFrequencyNode){calculationPeriodFrequencyIsNull_ = false;}
   else{calculationPeriodFrequencyIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- calculationPeriodFrequencyNode , address : " << calculationPeriodFrequencyNode << std::endl;
   #endif
   if(calculationPeriodFrequencyNode)
   {
      if(calculationPeriodFrequencyNode->Attribute("href") || calculationPeriodFrequencyNode->Attribute("id"))
      {
          if(calculationPeriodFrequencyNode->Attribute("id"))
          {
             calculationPeriodFrequencyIDRef_ = calculationPeriodFrequencyNode->Attribute("id");
             calculationPeriodFrequency_ = boost::shared_ptr<CalculationPeriodFrequency>(new CalculationPeriodFrequency(calculationPeriodFrequencyNode));
             calculationPeriodFrequency_->setID(calculationPeriodFrequencyIDRef_);
             IDManager::instance().setID(calculationPeriodFrequencyIDRef_,calculationPeriodFrequency_);
          }
          else if(calculationPeriodFrequencyNode->Attribute("href")) { calculationPeriodFrequencyIDRef_ = calculationPeriodFrequencyNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { calculationPeriodFrequency_ = boost::shared_ptr<CalculationPeriodFrequency>(new CalculationPeriodFrequency(calculationPeriodFrequencyNode));}
   }

   //calculationPeriodDatesAdjustmentsNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* calculationPeriodDatesAdjustmentsNode = xmlNode->FirstChildElement("calculationPeriodDatesAdjustments");

   if(calculationPeriodDatesAdjustmentsNode){calculationPeriodDatesAdjustmentsIsNull_ = false;}
   else{calculationPeriodDatesAdjustmentsIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- calculationPeriodDatesAdjustmentsNode , address : " << calculationPeriodDatesAdjustmentsNode << std::endl;
   #endif
   if(calculationPeriodDatesAdjustmentsNode)
   {
      if(calculationPeriodDatesAdjustmentsNode->Attribute("href") || calculationPeriodDatesAdjustmentsNode->Attribute("id"))
      {
          if(calculationPeriodDatesAdjustmentsNode->Attribute("id"))
          {
             calculationPeriodDatesAdjustmentsIDRef_ = calculationPeriodDatesAdjustmentsNode->Attribute("id");
             calculationPeriodDatesAdjustments_ = boost::shared_ptr<BusinessDayAdjustments>(new BusinessDayAdjustments(calculationPeriodDatesAdjustmentsNode));
             calculationPeriodDatesAdjustments_->setID(calculationPeriodDatesAdjustmentsIDRef_);
             IDManager::instance().setID(calculationPeriodDatesAdjustmentsIDRef_,calculationPeriodDatesAdjustments_);
          }
          else if(calculationPeriodDatesAdjustmentsNode->Attribute("href")) { calculationPeriodDatesAdjustmentsIDRef_ = calculationPeriodDatesAdjustmentsNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { calculationPeriodDatesAdjustments_ = boost::shared_ptr<BusinessDayAdjustments>(new BusinessDayAdjustments(calculationPeriodDatesAdjustmentsNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<AdjustableOrRelativeDate> PeriodicDates::getCalculationStartDate()
{
   if(!this->calculationStartDateIsNull_){
        if(calculationStartDateIDRef_ != ""){
             return boost::shared_static_cast<AdjustableOrRelativeDate>(IDManager::instance().getID(calculationStartDateIDRef_));
        }else{
             return this->calculationStartDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AdjustableOrRelativeDate>();
   }
}
boost::shared_ptr<AdjustableOrRelativeDate> PeriodicDates::getCalculationEndDate()
{
   if(!this->calculationEndDateIsNull_){
        if(calculationEndDateIDRef_ != ""){
             return boost::shared_static_cast<AdjustableOrRelativeDate>(IDManager::instance().getID(calculationEndDateIDRef_));
        }else{
             return this->calculationEndDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AdjustableOrRelativeDate>();
   }
}
boost::shared_ptr<CalculationPeriodFrequency> PeriodicDates::getCalculationPeriodFrequency()
{
   if(!this->calculationPeriodFrequencyIsNull_){
        if(calculationPeriodFrequencyIDRef_ != ""){
             return boost::shared_static_cast<CalculationPeriodFrequency>(IDManager::instance().getID(calculationPeriodFrequencyIDRef_));
        }else{
             return this->calculationPeriodFrequency_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CalculationPeriodFrequency>();
   }
}
boost::shared_ptr<BusinessDayAdjustments> PeriodicDates::getCalculationPeriodDatesAdjustments()
{
   if(!this->calculationPeriodDatesAdjustmentsIsNull_){
        if(calculationPeriodDatesAdjustmentsIDRef_ != ""){
             return boost::shared_static_cast<BusinessDayAdjustments>(IDManager::instance().getID(calculationPeriodDatesAdjustmentsIDRef_));
        }else{
             return this->calculationPeriodDatesAdjustments_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<BusinessDayAdjustments>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-shared-5-4/PeriodicDates.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class PeriodicDates : ISerialized
    {
        public PeriodicDates(XmlNode xmlNode)
        {
            XmlNode calculationStartDateNode = xmlNode.SelectSingleNode("calculationStartDate");
            
            if (calculationStartDateNode != null)
            {
                if (calculationStartDateNode.Attributes["href"] != null || calculationStartDateNode.Attributes["id"] != null) 
                {
                    if (calculationStartDateNode.Attributes["id"] != null) 
                    {
                        calculationStartDateIDRef_ = calculationStartDateNode.Attributes["id"].Value;
                        AdjustableOrRelativeDate ob = new AdjustableOrRelativeDate(calculationStartDateNode);
                        IDManager.SetID(calculationStartDateIDRef_, ob);
                    }
                    else if (calculationStartDateNode.Attributes["href"] != null)
                    {
                        calculationStartDateIDRef_ = calculationStartDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        calculationStartDate_ = new AdjustableOrRelativeDate(calculationStartDateNode);
                    }
                }
                else
                {
                    calculationStartDate_ = new AdjustableOrRelativeDate(calculationStartDateNode);
                }
            }
            
        
            XmlNode calculationEndDateNode = xmlNode.SelectSingleNode("calculationEndDate");
            
            if (calculationEndDateNode != null)
            {
                if (calculationEndDateNode.Attributes["href"] != null || calculationEndDateNode.Attributes["id"] != null) 
                {
                    if (calculationEndDateNode.Attributes["id"] != null) 
                    {
                        calculationEndDateIDRef_ = calculationEndDateNode.Attributes["id"].Value;
                        AdjustableOrRelativeDate ob = new AdjustableOrRelativeDate(calculationEndDateNode);
                        IDManager.SetID(calculationEndDateIDRef_, ob);
                    }
                    else if (calculationEndDateNode.Attributes["href"] != null)
                    {
                        calculationEndDateIDRef_ = calculationEndDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        calculationEndDate_ = new AdjustableOrRelativeDate(calculationEndDateNode);
                    }
                }
                else
                {
                    calculationEndDate_ = new AdjustableOrRelativeDate(calculationEndDateNode);
                }
            }
            
        
            XmlNode calculationPeriodFrequencyNode = xmlNode.SelectSingleNode("calculationPeriodFrequency");
            
            if (calculationPeriodFrequencyNode != null)
            {
                if (calculationPeriodFrequencyNode.Attributes["href"] != null || calculationPeriodFrequencyNode.Attributes["id"] != null) 
                {
                    if (calculationPeriodFrequencyNode.Attributes["id"] != null) 
                    {
                        calculationPeriodFrequencyIDRef_ = calculationPeriodFrequencyNode.Attributes["id"].Value;
                        CalculationPeriodFrequency ob = new CalculationPeriodFrequency(calculationPeriodFrequencyNode);
                        IDManager.SetID(calculationPeriodFrequencyIDRef_, ob);
                    }
                    else if (calculationPeriodFrequencyNode.Attributes["href"] != null)
                    {
                        calculationPeriodFrequencyIDRef_ = calculationPeriodFrequencyNode.Attributes["href"].Value;
                    }
                    else
                    {
                        calculationPeriodFrequency_ = new CalculationPeriodFrequency(calculationPeriodFrequencyNode);
                    }
                }
                else
                {
                    calculationPeriodFrequency_ = new CalculationPeriodFrequency(calculationPeriodFrequencyNode);
                }
            }
            
        
            XmlNode calculationPeriodDatesAdjustmentsNode = xmlNode.SelectSingleNode("calculationPeriodDatesAdjustments");
            
            if (calculationPeriodDatesAdjustmentsNode != null)
            {
                if (calculationPeriodDatesAdjustmentsNode.Attributes["href"] != null || calculationPeriodDatesAdjustmentsNode.Attributes["id"] != null) 
                {
                    if (calculationPeriodDatesAdjustmentsNode.Attributes["id"] != null) 
                    {
                        calculationPeriodDatesAdjustmentsIDRef_ = calculationPeriodDatesAdjustmentsNode.Attributes["id"].Value;
                        BusinessDayAdjustments ob = new BusinessDayAdjustments(calculationPeriodDatesAdjustmentsNode);
                        IDManager.SetID(calculationPeriodDatesAdjustmentsIDRef_, ob);
                    }
                    else if (calculationPeriodDatesAdjustmentsNode.Attributes["href"] != null)
                    {
                        calculationPeriodDatesAdjustmentsIDRef_ = calculationPeriodDatesAdjustmentsNode.Attributes["href"].Value;
                    }
                    else
                    {
                        calculationPeriodDatesAdjustments_ = new BusinessDayAdjustments(calculationPeriodDatesAdjustmentsNode);
                    }
                }
                else
                {
                    calculationPeriodDatesAdjustments_ = new BusinessDayAdjustments(calculationPeriodDatesAdjustmentsNode);
                }
            }
            
        
        }
        
    
        #region CalculationStartDate_
        private AdjustableOrRelativeDate calculationStartDate_;
        public AdjustableOrRelativeDate CalculationStartDate_
        {
            get
            {
                if (this.calculationStartDate_ != null)
                {
                    return this.calculationStartDate_; 
                }
                else if (this.calculationStartDateIDRef_ != null)
                {
                    calculationStartDate_ = IDManager.getID(calculationStartDateIDRef_) as AdjustableOrRelativeDate;
                    return this.calculationStartDate_; 
                }
                else
                {
                      return this.calculationStartDate_; 
                }
            }
            set
            {
                if (this.calculationStartDate_ != value)
                {
                    this.calculationStartDate_ = value;
                }
            }
        }
        #endregion
        
        public string calculationStartDateIDRef_ { get; set; }
        #region CalculationEndDate_
        private AdjustableOrRelativeDate calculationEndDate_;
        public AdjustableOrRelativeDate CalculationEndDate_
        {
            get
            {
                if (this.calculationEndDate_ != null)
                {
                    return this.calculationEndDate_; 
                }
                else if (this.calculationEndDateIDRef_ != null)
                {
                    calculationEndDate_ = IDManager.getID(calculationEndDateIDRef_) as AdjustableOrRelativeDate;
                    return this.calculationEndDate_; 
                }
                else
                {
                      return this.calculationEndDate_; 
                }
            }
            set
            {
                if (this.calculationEndDate_ != value)
                {
                    this.calculationEndDate_ = value;
                }
            }
        }
        #endregion
        
        public string calculationEndDateIDRef_ { get; set; }
        #region CalculationPeriodFrequency_
        private CalculationPeriodFrequency calculationPeriodFrequency_;
        public CalculationPeriodFrequency CalculationPeriodFrequency_
        {
            get
            {
                if (this.calculationPeriodFrequency_ != null)
                {
                    return this.calculationPeriodFrequency_; 
                }
                else if (this.calculationPeriodFrequencyIDRef_ != null)
                {
                    calculationPeriodFrequency_ = IDManager.getID(calculationPeriodFrequencyIDRef_) as CalculationPeriodFrequency;
                    return this.calculationPeriodFrequency_; 
                }
                else
                {
                      return this.calculationPeriodFrequency_; 
                }
            }
            set
            {
                if (this.calculationPeriodFrequency_ != value)
                {
                    this.calculationPeriodFrequency_ = value;
                }
            }
        }
        #endregion
        
        public string calculationPeriodFrequencyIDRef_ { get; set; }
        #region CalculationPeriodDatesAdjustments_
        private BusinessDayAdjustments calculationPeriodDatesAdjustments_;
        public BusinessDayAdjustments CalculationPeriodDatesAdjustments_
        {
            get
            {
                if (this.calculationPeriodDatesAdjustments_ != null)
                {
                    return this.calculationPeriodDatesAdjustments_; 
                }
                else if (this.calculationPeriodDatesAdjustmentsIDRef_ != null)
                {
                    calculationPeriodDatesAdjustments_ = IDManager.getID(calculationPeriodDatesAdjustmentsIDRef_) as BusinessDayAdjustments;
                    return this.calculationPeriodDatesAdjustments_; 
                }
                else
                {
                      return this.calculationPeriodDatesAdjustments_; 
                }
            }
            set
            {
                if (this.calculationPeriodDatesAdjustments_ != value)
                {
                    this.calculationPeriodDatesAdjustments_ = value;
                }
            }
        }
        #endregion
        
        public string calculationPeriodDatesAdjustmentsIDRef_ { get; set; }
        
    
        
    
    }
    
}

